compiler/noirc_frontend/src/elaborator/unquote.rs

```diff
@@ -1,306 +1,41 @@
-            }
-            ExpressionKind::Call(call) => self.find_unquoted_exprs_in_call(call, unquoted_exprs),
-            ExpressionKind::MethodCall(call) => {
-                self.find_unquoted_exprs_in_method_call(call, unquoted_exprs);
-            }
-            ExpressionKind::Constructor(constructor) => {
-                self.find_unquoted_exprs_in_constructor(constructor, unquoted_exprs);
-            }
-            ExpressionKind::MemberAccess(access) => {
-                self.find_unquoted_exprs_in_access(access, unquoted_exprs);
-            }
-            ExpressionKind::Cast(cast) => self.find_unquoted_exprs_in_cast(cast, unquoted_exprs),
-            ExpressionKind::Infix(infix) => {
-                self.find_unquoted_exprs_in_infix(infix, unquoted_exprs);
-            }
-            ExpressionKind::If(if_) => self.find_unquoted_exprs_in_if(if_, unquoted_exprs),
-            ExpressionKind::Variable(_, _) => (),
-            ExpressionKind::Tuple(tuple) => {
-                self.find_unquoted_exprs_in_tuple(tuple, unquoted_exprs);
-            }
-            ExpressionKind::Lambda(lambda) => {
-                self.find_unquoted_exprs_in_lambda(lambda, unquoted_exprs);
-            }
-            ExpressionKind::Parenthesized(expr) => {
-                self.find_unquoted_exprs_in_expr(expr, unquoted_exprs);
-            }
-            ExpressionKind::Quote(quote, _) => {
-                self.find_unquoted_exprs_in_block(quote, unquoted_exprs);
-            }
-            ExpressionKind::Comptime(block, _) => {
-                self.find_unquoted_exprs_in_block(block, unquoted_exprs);
-            }
-            ExpressionKind::Resolved(_) => (),
-            ExpressionKind::Error => (),
-            ExpressionKind::UnquoteMarker(_) => (),
-            ExpressionKind::Unquote(unquoted) => {
-                // Avoid an expensive clone for unquoted
-                let empty_expr = Expression::new(ExpressionKind::Error, unquoted.span);
-                let unquote = std::mem::replace(unquoted.as_mut(), empty_expr);
-                self.replace_unquote(expr, unquote, unquoted_exprs);
-            }
-        }
-    }
-
-    fn find_unquoted_exprs_in_literal(
-        &mut self,
-        literal: &mut Literal,
-        unquoted_exprs: &mut Vec<ExprId>,
-    ) {
-        match literal {
-            Literal::Array(array) | Literal::Slice(array) => match array {
-                ArrayLiteral::Standard(elements) => {
-                    for element in elements {
-                        self.find_unquoted_exprs_in_expr(element, unquoted_exprs);
+    /// Go through the given tokens looking for a '$' token followed by a variable to unquote.
+    /// Each time these two tokens are found, they are replaced by a new UnquoteMarker token
+    /// containing the ExprId of the resolved variable to unquote.
+    pub fn find_unquoted_exprs_tokens(&mut self, tokens: Tokens) -> Tokens {
```

Comment:
Invalid unquotes are just left in the token stream since this will only remove `$ ident` pairs. Token streams are allowed to be temporarily invalid so we don't issue an errors here. We'd wait until the macro is unquoted when it is parsed and the parser would issue an error. That's when we issue the new `InterpreterError::FailedToParseMacro` error.